File: src/main/tracker.ts
import { existsSync, readFileSync } from 'node:fs'
import { decryptAndParseES3, Es3DecryptError } from './es3'
import { GoldFlowTracker, type GoldFlowState } from './goldFlow'
import { HeroEventsTracker, type HeroEventsState } from './heroEvents'
import { flushHistory, loadHistory, saveHistory } from './history'
import { locateSave } from './locator'
import { parseSnapshot } from './parser'
import { StageEventsTracker, type StageEventsState } from './stageEvents'
import { StageFarmTracker, type StageFarmState } from './stageFarm'
import * as store from './store'
import { SaveWatcher } from './watcher'
import type { TrackerState } from '@shared/types'

// Intervalo do heartbeat (A2): re-emite o estado mesmo sem mudança no save, para
// provar que o tracker está vivo e atualizar "tempo desde a última mudança".
const HEARTBEAT_MS = 5000

export class Tracker {
  private watcher: SaveWatcher | null = null
  private goldFlow = new GoldFlowTracker()
  private heroEvents = new HeroEventsTracker()
  private stageEvents = new StageEventsTracker()
  private stageFarm = new StageFarmTracker()
  private heartbeatTimer: NodeJS.Timeout | null = null
  private trackedPath: string | null = null
  private state: TrackerState = {
    status: 'no-save',
    savePath: null,
    hasKey: false,
    lastError: null,
    snapshot: null,
    lastChangeAt: null,
    heartbeatAt: null
  }

  constructor(private readonly onUpdate: (state: TrackerState) => void) {}

  getState(): TrackerState {
    return this.state
  }

  /** (Re)avalia caminho do save, inicia o watcher e faz uma leitura. */
  start(): void {
    const savePath = store.getSavePathOverride() ?? locateSave()
    // Trocar de arquivo de save recarrega o histórico persistido daquele save (I6):
    // o estado é isolado por caminho, então cada save retoma seus próprios eventos.
    if (savePath !== this.trackedPath) {
      this.goldFlow.restore(loadHistory<GoldFlowState>(savePath, 'goldFlow'))
      this.heroEvents.restore(loadHistory<HeroEventsState>(savePath, 'heroEvents'))
      this.stageEvents.restore(loadHistory<StageEventsState>(savePath, 'stageEvents'))
      this.stageFarm.restore(loadHistory<StageFarmState>(savePath, 'stageFarm'))
      this.trackedPath = savePath
    }
    this.state.savePath = savePath
    this.state.hasKey = store.hasKey()

    if (this.watcher) {
      this.watcher.stop()
      this.watcher = null
    }

    this.startHeartbeat()

    if (savePath && existsSync(savePath)) {
      this.watcher = new SaveWatcher(savePath, () => this.readSave())
      this.watcher.start()
    } else {
      this.state.status = 'no-save'
      this.state.snapshot = null
      this.emit()
    }
  }

  /** Heartbeat (A2): pulso periódico que re-emite o estado mesmo sem mudança no save. */
  private startHeartbeat(): void {
    if (this.heartbeatTimer) clearInterval(this.heartbeatTimer)
    this.heartbeatTimer = setInterval(() => {
      this.state.heartbeatAt = Date.now()
      this.emit()
    }, HEARTBEAT_MS)
  }

  private readSave(): void {
    const savePath = this.state.savePath
    if (!savePath || !existsSync(savePath)) {
      this.update({ status: 'no-save', snapshot: null, lastError: null })
      return
    }

    const key = store.getKey()
    this.state.hasKey = key !== null
    if (!key) {
      this.update({ status: 'no-key', lastError: null })
      return
    }

    try {
      const buffer = readFileSync(savePath)
      const json = decryptAndParseES3(buffer, key)
      // I7: o snapshot NÃO carrega o JSON bruto (evita serializar o save inteiro a cada
      // leitura). O visualizador de calibração busca o raw sob demanda via readRawSave().
      const snapshot = parseSnapshot(json)
      const flow = this.goldFlow.record(snapshot.capturedAt, snapshot.gold)
      if (flow) snapshot.goldFlow = flow
      const heroEvents = this.heroEvents.record(snapshot.capturedAt, snapshot.heroes)
      if (heroEvents) snapshot.heroEvents = heroEvents
      const stageEvents = this.stageEvents.record(
        snapshot.capturedAt,
        snapshot.stage,
        snapshot.maxCompletedStage
      )
      if (stageEvents) snapshot.stageEvents = stageEvents
      // XP da conta = soma de HeroExp (proxy; reseta em level-up, mas deltas negativos
      // são ignorados pelo tracker — ver stageFarm.ts).
      const totalExp = snapshot.heroes.length
        ? snapshot.heroes.reduce((sum, h) => sum + (h.exp ?? 0), 0)
        : null
      const stageFarm = this.stageFarm.record(
        snapshot.capturedAt,
        snapshot.stage?.raw ?? null,
        snapshot.gold,
        totalExp,
        snapshot.totalKills
      )
      if (stageFarm) snapshot.stageFarm = stageFarm
      // Persiste o histórico desta leitura (I6, debounced por save).
      saveHistory(savePath, 'goldFlow', this.goldFlow.serialize())
      saveHistory(savePath, 'heroEvents', this.heroEvents.serialize())
      saveHistory(savePath, 'stageEvents', this.stageEvents.serialize())
      saveHistory(savePath, 'stageFarm', this.stageFarm.serialize())
      const now = Date.now()
      this.update({
        status: 'monitoring',
        snapshot,
        lastError: null,
        lastChangeAt: now,
        heartbeatAt: now
      })
    } catch (err) {
      const message =
        err instanceof Es3DecryptError
          ? err.message
          : err instanceof Error
            ? err.message
            : 'Erro desconhecido ao ler o save.'
      this.update({ status: 'error', lastError: message })
    }
  }

  refresh(): TrackerState {
    this.start()
    return this.state
  }

  /**
   * Lê o save agora e devolve só o JSON bruto do player (I7) — sob demanda, para o
   * visualizador de calibração. Não toca no estado/snapshot em curso. Retorna null se
   * não houver save/chave ou em caso de erro de leitura.
   */
  readRawSave(): unknown | null {
    const savePath = this.state.savePath
    if (!savePath || !existsSync(savePath)) return null
    const key = store.getKey()
    if (!key) return null
    try {
      const buffer = readFileSync(savePath)
      const json = decryptAndParseES3(buffer, key)
      return parseSnapshot(json, true).raw ?? null
    } catch {
      return null
    }
  }

  private update(patch: Partial<TrackerState>): void {
    this.state = { ...this.state, ...patch }
    this.emit()
  }

  private emit(): void {
    this.onUpdate(this.state)
  }

  stop(): void {
    this.watcher?.stop()
    this.watcher = null
    if (this.heartbeatTimer) {
      clearInterval(this.heartbeatTimer)
      this.heartbeatTimer = null
    }
    // Garante que o histórico pendente (debounced) vá pro disco antes de fechar.
    flushHistory()
  }
}

File: dotnet/TbhTracker.App/Services/Tracker.cs
using System.Text.Json;
using TbhTracker.Core;
using TbhTracker.Core.Logic;
using TbhTracker.Core.Models;

namespace TbhTracker.App.Services;

/// <summary>Port de src/main/tracker.ts. Orquestra leitura passiva do save: localiza,
/// observa (watcher), descriptografa+parseia, alimenta os trackers de fluxo/eventos,
/// persiste o historico e emite TrackerState. Heartbeat re-emite o estado a cada 5s.</summary>
public sealed class Tracker : IDisposable
{
    private const int HeartbeatMs = 5000;

    private readonly ConfigStore _store;
    private readonly HistoryStore _history;
    private readonly Locator _locator;

    private readonly GoldFlowTracker _goldFlow = new();
    private readonly HeroEventsTracker _heroEvents = new();
    private readonly StageEventsTracker _stageEvents = new();
    private readonly StageFarmTracker _stageFarm = new();

    private SaveWatcher? _watcher;
    private Timer? _heartbeatTimer;
    private string? _trackedPath;
    private readonly Lock _gate = new();

    private TrackerState _state = new()
    {
        Status = ConnectionStatus.NoSave,
        SavePath = null,
        HasKey = false,
        LastError = null,
        Snapshot = null,
        LastChangeAt = null,
        HeartbeatAt = null
    };

    /// <summary>Emitido sempre que o estado muda (ou no heartbeat).</summary>
    public event Action<TrackerState>? OnState;

    public Tracker(ConfigStore store, HistoryStore history, Locator locator)
    {
        _store = store;
        _history = history;
        _locator = locator;
    }

    public TrackerState GetState() => _state;

    private static long NowMs() => DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

    /// <summary>(Re)avalia caminho do save, inicia o watcher e faz uma leitura.</summary>
    public void Start()
    {
        var savePath = _store.GetSavePathOverride() ?? _locator.LocateSave();
        if (savePath != _trackedPath)
        {
            _goldFlow.Restore(_history.LoadHistory<GoldFlowState>(savePath, "goldFlow"));
            _heroEvents.Restore(_history.LoadHistory<HeroEventsState>(savePath, "heroEvents"));
            _stageEvents.Restore(_history.LoadHistory<StageEventsState>(savePath, "stageEvents"));
            _stageFarm.Restore(_history.LoadHistory<StageFarmState>(savePath, "stageFarm"));
            _trackedPath = savePath;
        }
        _state.SavePath = savePath;
        _state.HasKey = _store.HasKey();

        _watcher?.Stop();
        _watcher = null;

        StartHeartbeat();

        if (!string.IsNullOrEmpty(savePath) && File.Exists(savePath))
        {
            _watcher = new SaveWatcher(savePath, ReadSave);
            _watcher.Start();
        }
        else
        {
            _state.Status = ConnectionStatus.NoSave;
            _state.Snapshot = null;
            Emit();
        }
    }

    private void StartHeartbeat()
    {
        _heartbeatTimer?.Dispose();
        _heartbeatTimer = new Timer(_ =>
        {
            _state.HeartbeatAt = NowMs();
            Emit();
        }, null, HeartbeatMs, HeartbeatMs);
    }

    private void ReadSave()
    {
        lock (_gate)
        {
            var savePath = _state.SavePath;
            if (string.IsNullOrEmpty(savePath) || !File.Exists(savePath))
            {
                Update(s => { s.Status = ConnectionStatus.NoSave; s.Snapshot = null; s.LastError = null; });
                return;
            }

            var key = _store.GetKey();
            _state.HasKey = key != null;
            if (key == null)
            {
                Update(s => { s.Status = ConnectionStatus.NoKey; s.LastError = null; });
                return;
            }

            try
            {
                var buffer = File.ReadAllBytes(savePath);
                var json = Es3Crypto.DecryptAndParse(buffer, key);
                var snapshot = SaveParser.ParseSnapshot(json);

                var flow = _goldFlow.Record(snapshot.CapturedAt, snapshot.Gold);
                if (flow != null) snapshot.GoldFlow = flow;

                var heroEvents = _heroEvents.Record(snapshot.CapturedAt, snapshot.Heroes);
                if (heroEvents != null) snapshot.HeroEvents = heroEvents;

                var stageEvents = _stageEvents.Record(snapshot.CapturedAt, snapshot.Stage, snapshot.MaxCompletedStage);
                if (stageEvents != null) snapshot.StageEvents = stageEvents;

                double? totalExp = snapshot.Heroes.Count > 0
                    ? snapshot.Heroes.Sum(h => h.Exp ?? 0)
                    : null;
                var stageFarm = _stageFarm.Record(
                    snapshot.CapturedAt,
                    snapshot.Stage?.Raw,
                    snapshot.Gold,
                    totalExp,
                    snapshot.TotalKills,
                    snapshot.ArrangedHeroKeys);
                if (stageFarm != null) snapshot.StageFarm = stageFarm;

                _history.SaveHistory(savePath, "goldFlow", _goldFlow.Serialize());
                _history.SaveHistory(savePath, "heroEvents", _heroEvents.Serialize());
                _history.SaveHistory(savePath, "stageEvents", _stageEvents.Serialize());
                _history.SaveHistory(savePath, "stageFarm", _stageFarm.Serialize());

                var now = NowMs();
                Update(s =>
                {
                    s.Status = ConnectionStatus.Monitoring;
                    s.Snapshot = snapshot;
                    s.LastError = null;
                    s.LastChangeAt = now;
                    s.HeartbeatAt = now;
                });
            }
            catch (Exception err)
            {
                Update(s => { s.Status = ConnectionStatus.Error; s.LastError = err.Message; });
            }
        }
    }

    public TrackerState Refresh()
    {
        Start();
        return _state;
    }

    /// <summary>Apaga o historico de medicoes de farm (memoria + base persistida) do save atual.</summary>
    public TrackerState ClearStageFarmHistory()
    {
        lock (_gate)
        {
            _stageFarm.Reset();
            _history.SaveHistory(_state.SavePath, "stageFarm", _stageFarm.Serialize());
            _history.Flush();
            if (_state.Snapshot != null) _state.Snapshot.StageFarm = null;
            Emit();
        }
        return _state;
    }

    /// <summary>Le o save agora e devolve so o JSON bruto do player (sob demanda, para o
    /// visualizador de calibracao). Nao toca no estado/snapshot em curso.</summary>
    public object? ReadRawSave()
    {
        var savePath = _state.SavePath;
        if (string.IsNullOrEmpty(savePath) || !File.Exists(savePath)) return null;
        var key = _store.GetKey();
        if (key == null) return null;
        try
        {
            var buffer = File.ReadAllBytes(savePath);
            var json = Es3Crypto.DecryptAndParse(buffer, key);
            return SaveParser.ParseSnapshot(json, includeRaw: true).Raw;
        }
        catch
        {
            return null;
        }
    }

    private void Update(Action<TrackerState> patch)
    {
        patch(_state);
        Emit();
    }

    private void Emit() => OnState?.Invoke(_state);

    public void Stop()
    {
        _watcher?.Stop();
        _watcher = null;
        _heartbeatTimer?.Dispose();
        _heartbeatTimer = null;
        _history.Flush();
    }

    public void Dispose() => Stop();
}
